### PR TITLE
Common client code and few bug fixes

### PR DESCRIPTION
**Pull request by hpya93**
Hi @jhand2 

Have added
- **Files: abi.go, client.go**: Code to issue RotateContextHandle, DeriveChild, ExtendTCI and Sign commands
- **File- sign.rs**: Bug fix for IS_SYMMETRIC flag in sign.rs file
- **File- sign.rs**: Bug fix  for DPE error code in sign.rs. It must be "ArgumentNotSupported" instead of "InvalidArgument" when DPE profile does not support a command flag.
- **File - CertifyKey.go, TagTci.go**:  Bug fix, the context handle created for Simulation mode is not used in code. The contexthandle is stored in variable "handle", but variable "ctx" is used
- **File - InitializeContext.go**: Separated out code for simulated context handle creation and put it as a func in InitializeContext.go file. This block was repeated many times.
-  **Files - negativeCases.go, simulator.go, verification.go**: Added a negativeCases.go - it will test InvalidHandle, wrongLocality, Unsupported DPE command, Unsupported command flags. Tried to address your comment in TagTCI PR by writing a separate function for checking WrongLocality and other negative cases..   

When this is merged I shall change other PRs in tune to the main branch with these changes. Kindly let us know the comments.

### CODE REVIEW

File: verification/abi.go [hunk 1]
```diff
@@ -3,11 +3,12 @@
 package verification
 
 import (
-	"errors"
 	"fmt"
 	"reflect"
 )
 
+var DefaultContextHandle = ContextHandle{0}
+
```
Comment (hpya93) [inline verification/abi.go:11]: Couldnt declare as constant as arrays are not allowed to be constants.

File: verification/certifyKey.go [hunk 2]
```diff
@@ -431,26 +431,10 @@ func checkCertificateStructure(t *testing.T, certBytes []byte) *x509.Certificate
...
-		} else {
-			t.Errorf("[ERROR]:  DPE instance doesn't support simulation contexts.")
-		}
-	} else {
-		//default context
-		ctx = [16]byte{0}
+func testCertifyKey(d TestDPEInstance, client DPEClient, t *testing.T, simulation bool) {
+	ctx := getContextHandle(d, client, t, simulation)
+	if simulation {
+		defer client.DestroyContext(ctx, 0)
```
Comment (hpya93) [inline verification/certifyKey.go:437]: Have put  this code block as a function, since it is getting repeated for each command using Simulation mode.

File: verification/negativeCases.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
+// Licensed under the Apache-2.0 license
```
Comment (hpya93) [inline verification/negativeCases.go:1]: File that tests WrongLoclaity, invalid handle, Unsupported command and UnsupportedCommand flags

File: verification/certifyKey.go [hunk 4]
```diff
@@ -500,6 +484,11 @@ func testCertifyKey(d TestDPEInstance, client DPEClient, t *testing.T, use_simul
 		// This also checks certificate lifetime, signatures as part of cert chain validation
 		validateLeafCertChain(t, certChain, leafCert)
 
+		// Reassign handle for simulation mode.
+		// However, this does not impact in default mode because
+		// same default context handle is returned in default mode.
+		ctx = &certifyKeyResp.Handle
+
```
Comment (hpya93) [inline verification/certifyKey.go:491]: This is required, else SimulationMode test fails after properly assigning simulation context handle

File: verification/tagTCI.go [hunk 5]
```diff
@@ -8,16 +8,14 @@ import (
...
-			t.Fatalf("Failed to initialize default context: %v", err)
-		}
-		defer client.DestroyContext(handle, DestroyDescendants)
+	testTagTCI(d, client, t, false)
+}
+
+func testTagTCI(d TestDPEInstance, client DPEClient, t *testing.T, simulation bool) {
+	ctx := getContextHandle(d, client, t, simulation)
+	if simulation {
+		defer client.DestroyContext(ctx, DestroyDescendants)
```
Comment (hpya93) [inline verification/tagTCI.go:18]: same problem here, handle is initialiaed but never used, also its scope is only within the IF block

File: verification/verification.go [hunk 6]
```diff
@@ -39,6 +41,18 @@ var TagTCITestCase = TestCase{
 var GetProfileTestCase = TestCase{
 	"GetProfile", TestGetProfile, []string{},
 }
+var InvalidHandleTestCase = TestCase{
+	"CheckInvalidHandle", TestInvalidHandle, []string{"Simulation", "RotateContext", "ExtendTci", "Tagging"},
+}
+var WrongLocalityTestCase = TestCase{
+	"CheckWrongLocality", TestWrongLocality, []string{"AutoInit", "RotateContext", "ExtendTci", "Tagging"},
+}
+var UnsupportedCommand = TestCase{
```
Comment (hpya93) [inline verification/verification.go:50]: Have added limited DPE support to test failure

File: dpe/src/commands/sign.rs [hunk 7]
```diff
@@ -15,7 +15,7 @@ pub struct SignFlags(u32);
 
 bitflags! {
     impl SignFlags: u32 {
-        const IS_SYMMETRIC = 1u32 << 31;
+        const IS_SYMMETRIC = 1u32 << 30;
```
Comment (hpya93) [inline dpe/src/commands/sign.rs:18]: A minor deviation from irot spec

File: dpe/src/commands/sign.rs [hunk 8]
```diff
@@ -83,7 +83,7 @@ impl CommandExecution for SignCmd {
     ) -> Result<Response, DpeErrorCode> {
         // Make sure the operation is supported.
         if !dpe.support.is_symmetric() && self.uses_symmetric() {
-            return Err(DpeErrorCode::InvalidArgument);
+            return Err(DpeErrorCode::ArgumentNotSupported);
```
Comment (hpya93) [inline dpe/src/commands/sign.rs:86]: All other commands return ArgumentNotSupported. Only Sign comamnd is returning a different error code, so changed it.

File: dpe/src/commands/sign.rs [hunk 9]
```diff
@@ -190,7 +190,7 @@ mod tests {
 
         // Bad argument
         assert_eq!(
-            Err(DpeErrorCode::InvalidArgument),
+            Err(DpeErrorCode::ArgumentNotSupported),
```
Comment (hpya93) [inline dpe/src/commands/sign.rs:193]: corresponding change added to rust testcase

File: verification/initializeContext.go [hunk 10]
```diff
@@ -67,3 +67,27 @@ func testInitContext(d TestDPEInstance, client DPEClient, t *testing.T, simulati
 		}
 	}
 }
+
+func getContextHandle(d TestDPEInstance, c DPEClient, t *testing.T, simulation bool) *ContextHandle {
```
Comment (jhand2) [inline verification/initializeContext.go:71]: nit: rename to `getInitialContextHandle`? Also would be good to have a function level comment explaining what this function is for.

File: verification/initializeContext.go [hunk 11]
```diff
@@ -67,3 +67,27 @@ func testInitContext(d TestDPEInstance, client DPEClient, t *testing.T, simulati
...
+	var handle *ContextHandle
+	var err error
+	if simulation {
+		if d.GetSupport().Simulation {
+			handle, err = c.InitializeContext(InitIsSimulation)
+			if err != nil {
+				t.Fatal("The instance should be able to create a simulation context.")
+			}
+			// Could prove difficult to prove it is a cryptographically secure random.
```
Comment (jhand2) [inline verification/initializeContext.go:80]: That's fine, I don't think it's necessary. Can remove this comment.

File: verification/negativeCases.go [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+var InvalidHandle = ContextHandle{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}
+
+func TestInvalidHandle(d TestDPEInstance, c DPEClient, t *testing.T) {
+	testInvalidHandle(d, c, t)
+}
+
+// Checks whether error is reported when non-existent handle is passed as input to DPE commands.
+// Exceptions are - GetProfile, InitializeContext, GetCertificateChain, GetTaggedTCI commands
+// which do not need context handle as input parameter.
+func testInvalidHandle(d TestDPEInstance, c DPEClient, t *testing.T) {
```
Comment (jhand2) [inline verification/negativeCases.go:19]: All this code can just be in the public function instead of calling a separate helper (same for all of these tests)

File: verification/tagTCI.go (hunk 5 above)
Comment (jhand2) [inline verification/tagTCI.go:18]: This doesn't need to be in a helper funciton

File: verification/verification.go [hunk 13]
```diff
@@ -39,6 +41,18 @@ var TagTCITestCase = TestCase{
...
+	"CheckSupportForCommand", TestUnsupportedCommand, []string{"AutoInit"},
+}
+var UnsupportedCommandFlag = TestCase{
+	"CheckSupportForCommmandFlag", TestUnsupportedCommandFlag, []string{"AutoInit", "RotateContext", "ExtendTci", "Tagging"},
+}
```
Comment (jhand2) [inline verification/verification.go:55]: Can you add all of these to the `AllTestCases` array?

File: verification/initializeContext.go (hunk 10 above)
Comment (hpya93) [reply verification/initializeContext.go:71]: Yes renamed and added a description.

File: verification/initializeContext.go (hunk 11 above)
Comment (hpya93) [reply verification/initializeContext.go:80]: This was there already in original code, but removed now.

File: verification/negativeCases.go (hunk 12 above)
Comment (hpya93) [reply verification/negativeCases.go:19]: Removed the helper, put the logic under public function

File: verification/tagTCI.go (hunk 5 above)
Comment (hpya93) [reply verification/tagTCI.go:18]: Removed the helper, put the logic under public function

File: verification/verification.go (hunk 13 above)
Comment (hpya93) [reply verification/verification.go:55]: **I have some problems with adding.**
1. The AllTestCases has a broad DPE suppport features, so it will not cause/catch errors in  UnsupportedCommand and UnsupportedCommandFlag  testcases, so left as it is.
2. However, I could add the Invalidhandle and Wrong locality test cases under AllTestCases without issues.

Comment (jhand2) [reply verification/verification.go:55]: I don't understand. AllTestCases is just a list if all the test cases. If the target doesn't have the support the test is requesting, the test will be skipped.

Comment (hpya93) [reply verification/verification.go:55]: Adding some more information for context:
1. AllTestCases is list of all testcases that is run with DefaultSupport test target which has more support features.
```
		{
			"DefaultSupport",
			getTestTarget([]string{"AutoInit", "Simulation", "X509", "IsCA", "Tagging", "RotateContext", "ExtendTci"}),
			AllTestCases,
		},
```
2. For **UnSupportedCommand testcase**, we **purposefully exclude** these support features to test for the "UnsupportedCommand" error code - especially **Tagging, RotateContext, ExtendTCI**
4. Similarly, for **UnSupportedCommandFlag testcase**, we **purposefully exclude** these support features to test for the "UnSupportedArgument" error code - especially **Simulation, IsCA, X509, ** 
5. Adding these tests in AllTestCases will ultimately test these two negative tests will Default support test target.
6. **If this happens, some cases in "UnSupportedCommand " and "UnSupportedCommandFlag" will fail as expected error will not be returned.**
7. Here is how it fails
```
    --- FAIL: TestRunAll/DefaultSupport-CheckSupportForCommand (0.00s)
    --- FAIL: TestRunAll/DefaultSupport-CheckSupportForCommmandFlag (0.01s)

```
- It gives many Error messages like so are given when testing ....
```
negativeCases.go:248: [ERROR]: Should return "argument is not supported by this profile, implementation, or integration", but returned no error

```

Comment (jhand2) [reply verification/verification.go:55]: Thanks for the explanation. I have opened https://github.com/chipsalliance/caliptra-dpe/issues/256 to track supporting this. We can merge this PR without this feature though.

**Review by jhand2 — APPROVED** [review]
